### PR TITLE
Adding examples into CMake build

### PR DESCRIPTION
**Pull request by Ali-Nasrolahi**
# CMake build

Good morning @aregtech and @biwiki, hope you're doing well.

I would like to present latest progress on CMake integration to `areg-sdk`.

## Change-logs

- Rest of examples from `10_locsvc` through `16_pubfsm` added to CMake build routine.
- Generated outputs would be part of `product` tree.
- Few improvements  on configurations happened for better solution.

## Next step

- At this point, I believe CMake could be a trustworthy build system for `areg-sdk` - naturally, after being tested and modified as you desired - and could replace Makefile buiild.
     - Of course, there's always some bugfixes and improvements to make, nevertheless with current state; we could move on for now and rest could be done by small bugfixes.

## Test

For testing result I have used following commands:

```bash
$ cmake -Bbuild # Creates build directory for cmake configs.
$ cmake --build build -j 4 # Builds areg-sdk core and examples by default.
```
Output files will stay in `product/<build-type>` directory tree.

### Note

Please test the build system and verify its integrity.
Unfortunately, I have not proper windows machine to test `Visual Studio` build system; therefore, it'll be nice if you could check it out.

As always open to any modification :)

### CODE REVIEW

File: examples/00_helloservice/generated/src/private/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+list(APPEND generated_SRC
```
Comment (biwiki) [inline examples/00_helloservice/generated/src/private/CMakeLists.txt]: Maybe `generated_SRC` can be named with a prefix, like `00_generated_SRC`. That way we dont have to unset the variables. Also having separated variables for each project can be good for debugging purpose

File: examples/00_helloservice/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+include(${00_helloservice_BASE}/twothreads/CMakeLists.txt)
+
+add_executable(${projName1} ${SRC1})
+add_executable(${projName2} ${SRC2})
+add_executable(${projName3} ${SRC3})
+add_executable(${projName4} ${SRC4})
+
+# Setting up appropriate flags
+setLibOpt(${gen_projName} ${generated_SRC})
+setopt(${projName1})
```
Comment (biwiki) [inline examples/00_helloservice/CMakeLists.txt]: Maybe try to use set**O**pt even when cmake accepts case insensitive here

File: examples/00_helloservice/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+setopt(${projName2})
+setopt(${projName3})
+setopt(${projName4})
+
+target_link_libraries(${projName1} ${gen_projName})
+target_link_libraries(${projName2} ${gen_projName})
+target_link_libraries(${projName3} ${gen_projName})
+target_link_libraries(${projName4} ${gen_projName})
```
Comment (biwiki) [inline examples/00_helloservice/CMakeLists.txt:38]: yeah, this is great

**Review by biwiki — APPROVED** [review]

**Comment by biwiki** [comment]
Looking great to me! :100:

File: examples/00_helloservice/generated/src/private/CMakeLists.txt (hunk 1 above)
Comment (Ali-Nasrolahi) [reply examples/00_helloservice/generated/src/private/CMakeLists.txt]: Resolved.

File: examples/00_helloservice/CMakeLists.txt (hunk 2 above)
Comment (Ali-Nasrolahi) [reply examples/00_helloservice/CMakeLists.txt]: Resolved.

**Review by aregtech — APPROVED** [review]
Thank you, @Ali-Nasrolahi 
Good work!

**Comment by aregtech** [comment]
@Ali-Nasrolahi, thank you very much for the updates! Please let me know whether there are planned changes to make. Soon I'll create a candidate branch to collect all changes and continue polishing :) 

@biwiki, many thanks for the review!

**Comment by Ali-Nasrolahi** [comment]
Well, just a branch for future patches will be good. :)

---
BTW, 
I would like to thank  you both @aregtech and @biwiki for the whole these months.
This project was/is my first official contribution to open source community and it felt great to work with you guys on it.
I greatly appreciate your patience through these couple months,  and hope for better and more collaboration in future.

**Pull request merged**